File: MVideo/ProductsService.py
import DataParser
from ProductsStorage import ProductsStorage


class ProductsService:

    __products_storage: ProductsStorage

    def __init__(self):
        input_lines = DataParser.parse_csv_file()
        self.__products_storage = ProductsStorage(input_lines)

    def get_recommended_products_ids(self, product_id: str):
        int_id = self.__products_storage.get_product_int_id(product_id)
        if int_id == -1:
            return list()

        result: list[str] = list()
        recommendations = self.__products_storage.products_recommendations[int_id]
        for recommendationsList in recommendations.values():
            for recommendation in recommendationsList:
                str_id = self.__products_storage.get_product_str_id(recommendation)
                result.append(str_id)
        return result

    def get_recommended_products_ids_with_min_rank(self, product_id: str, min_rank: float):
        int_id = self.__products_storage.get_product_int_id(product_id)
        if int_id == -1:
            return list()

        result: list[str] = list()
        recommendations = self.__products_storage.products_recommendations[int_id]
        for recommendation_rank in recommendations:
            if recommendation_rank >= min_rank:
                for recommendation in recommendations[recommendation_rank]:
                    str_id = self.__products_storage.get_product_str_id(recommendation)
                    result.append(str_id)
        return result

File: MVideo/c#/TestMVideo/ProductsService.cs
#region using

using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.Extensions.Configuration;

#endregion

namespace TestMVideo
{
    public class ProductsService
    {
        private readonly ProductsStorage productsStorage;

        public ProductsService(IConfiguration configuration)
        {
            var dataParser = new DataParser();
            var inputLines = dataParser.ParseCsvFile(configuration["DataFilePath"]);
            productsStorage = new ProductsStorage(inputLines);
        }

        public IEnumerable<string> GetRecommendedProductsIds(string productId)
        {
            var intId = productsStorage.GetProductsIntId(productId);
            return intId == -1
                ? Array.Empty<string>()
                : productsStorage.ProductsWithRecommendations[intId]
                    .SelectMany(x => x.Value)
                    .Select(x => productsStorage.GetProductsStringId(x));
        }

        public IEnumerable<string> GetRecommendedProductsIdsWithMinRank(string productId, float minRank)
        {
            var intId = productsStorage.GetProductsIntId(productId);
            return intId == -1
                ? Array.Empty<string>()
                : productsStorage.ProductsWithRecommendations[intId]
                    .Where(x => x.Key >= minRank)
                    .SelectMany(x => x.Value)
                    .Select(x => productsStorage.GetProductsStringId(x));
        }
    }
}